pkg/apis/storage/types.go

```diff
@@ -424,3 +436,67 @@ type CSINodeList struct {
 	// items is the list of CSINode
 	Items []CSINode
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIStorageCapacity stores the result of one CSI GetCapacity call for one
+// driver, one topology segment, and the parameters of one storage class.
+//
+// This is an alpha feature and only available when the CSIStorageCapacity feature is enabled.
+type CSIStorageCapacity struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning. It must be
+	// be a DNS subdomain (dots allowed, 253 characters). To ensure that
+	// there are no conflicts with other CSI drivers on the cluster, the recommendation
+	// is to use csisc-<uuid>, a generated name, or a reverse-domain name which ends
+	// with the unique CSI driver name.
+	//
+	// Objects are not namespaced.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// The CSI driver that provides the storage.
+	// This must be the string returned by the CSI GetPluginName() call.
+	// This field is immutable.
+	DriverName string
```

Comment:
Doesn't StorageClass already have this info?  Why do we need it here?  Is this just an opportunity for inconsistency?

Reply:
The storage class might get deleted before this object, in which case it becomes hard to determine which driver the object belongs to. The scheduler doesn't care (it can just compare the storage class name), but a human listing these objects may care and would benefit from having this field.

But even the scheduler benefits. There's an edge case: a storage class might get updated so that it references a different driver. Existing `CSIStorageCapacity` objects then have the right storage class name, but should better be ignored by the scheduler until eventually they'll get deleted by their owner. Comparing the driver name achieves that.

Reply:
Not just the scheduler. The producer also needs to know whether there are any obsolete objects lying around that it should delete. This can happen when the owner of the objects is the statefulset and the pod which runs external-provisioner gets restarted or the leader changes.

The producer can then list all objects and do a client-side filtering, but a field filter by driver name would be more efficient, right?

Reply:
> The storage class might get deleted before this object, in which case it becomes hard to determine which driver the object belongs to. 

The storage class parameters might change, too, but we don't copy those into here?

Reply:
True, the affect for the scheduler is the same ("this object is stale"). But the storage class parameters don't affect ownership of the `CSIStorageCapacity` object (the same provisioner as before will update the object) while driver name does (one provisioner will notice that it is no longer responsible for the storage class and delete the obsolete object, another provisioner will potentially create a new one).

Having said that, I am fine with removing the field. It's not required.

Reply:
@msau42 @saad-ali We have discussed this before wrt PV/PVC and we've not deemed it important enough to really try to prevent.  Right?  what about this case?

To my eyes it seems more likely that someone would change the `parameters` inside a StorageClass than the driver, which could have material impact on the meaning on the capacity.

Maybe I am over-blowing the scope of the badness, but it seems inconsistent.  I ACK concerns around relationship, but it feels like that should maybe be more formal (OwerRef)?

Reply:
>  I ACK concerns around relationship, but it feels like that should maybe be more formal (OwerRef)?

That's exactly how ownership is meant to be tracked, and therefore the driver name field is not really needed. Can we resolve the inconsistency by dropping the driver name? You have convinced me that it's not needed.

Regarding "badness": it is part of the overall design that information may be wrong and/or out-dated. That's what the recovery mechanisms are needed for. The "use information from some other driver" is just another variant of that.

Reply:
StorageClass parameters are immutable so to modify it you actually have to delete and create a new object with the same name.  I think the suggestion here was to maybe having an OwnerRef to StorageClass instead of just a name. So once the StorageClass is recreated, that invalidates the capacity object.

Reply:
>  I think the suggestion here was to have an OwnerRef to StorageClass instead of just a name.

That works technically and does indeed avoid the ambiguity whether the object really applies to the current version of the storage class. But should that then replace the existing field? Wouldn't it be rather unusual to require that an object must have a certain object reference?

A simpler solution would be to include a `StorageClassUID` field. Name + uid are enough to implement the "has the storage class been replaced" check. Shall I add that?